src/org/sosy_lab/java_smt/solvers/cvc5/CVC5InterpolatingProver.java

```diff
@@ -0,0 +1,370 @@
+//
+// SPDX-License-Identifier: Apache-2.0
+
+package org.sosy_lab.java_smt.solvers.cvc5;
+
+import com.google.common.base.Preconditions;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableSet;
+import com.google.common.collect.Iterables;
+import io.github.cvc5.Kind;
+import io.github.cvc5.Solver;
+import io.github.cvc5.Term;
+import java.util.Collection;
+import java.util.List;
+import java.util.Set;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+import org.sosy_lab.common.ShutdownNotifier;
+import org.sosy_lab.java_smt.api.BooleanFormula;
+import org.sosy_lab.java_smt.api.FormulaManager;
+import org.sosy_lab.java_smt.api.InterpolatingProverEnvironment;
+import org.sosy_lab.java_smt.api.SolverContext.ProverOptions;
+import org.sosy_lab.java_smt.api.SolverException;
+
+public class CVC5InterpolatingProver extends CVC5AbstractProver<Term>
+    implements InterpolatingProverEnvironment<Term> {
+
+  private final FormulaManager mgr;
+  private final Set<ProverOptions> solverOptions;
+  private final int seed;
+
+  CVC5InterpolatingProver(
+      CVC5FormulaCreator pFormulaCreator,
+      ShutdownNotifier pShutdownNotifier,
+      @SuppressWarnings("unused") int randomSeed,
+      Set<ProverOptions> pOptions,
+      FormulaManager pMgr) {
+    super(pFormulaCreator, pShutdownNotifier, randomSeed, pOptions, pMgr);
+    mgr = pMgr;
+    solverOptions = pOptions;
+    seed = randomSeed;
+  }
+
+  /**
+   * Sets the same solver Options of the Original Solver to the separate solvertoSet, except for
+   * produce-interpolants which is set here. From CVC5AbstractProver Line 66
+   */
+  private void setSolverOptions(int randomSeed, Set<ProverOptions> pOptions, Solver solvertoSet) {
+    if (incremental) {
+      solvertoSet.setOption("incremental", "true");
+    }
+    if (pOptions.contains(ProverOptions.GENERATE_MODELS)) {
+      solvertoSet.setOption("produce-models", "true");
+    }
+    if (pOptions.contains(ProverOptions.GENERATE_UNSAT_CORE)) {
+      solvertoSet.setOption("produce-unsat-cores", "true");
+    }
+    solvertoSet.setOption("produce-assertions", "true");
+    solvertoSet.setOption("dump-models", "true");
+    solvertoSet.setOption("output-language", "smt2");
+    solvertoSet.setOption("seed", String.valueOf(randomSeed));
+    solvertoSet.setOption("produce-interpolants", "true");
+
+    // Set Strings option to enable all String features (such as lessOrEquals)
+    solvertoSet.setOption("strings-exp", "true");
+
+    // Enable more complete quantifier solving (for more info see CVC5QuantifiedFormulaManager)
+    solvertoSet.setOption("full-saturate-quant", "true");
+  }
+
+  @Override
+  public Term addConstraint(BooleanFormula pConstraint) throws InterruptedException {
+    Preconditions.checkState(!closed);
+    Term t = creator.extractInfo(pConstraint);
+
+    super.addConstraint(pConstraint);
+    return t; // t is not returned in the Abstract Class
```

Comment:
Are you sure to return the internal reference?

Please consider the following cases:
- What happens when adding the same constraint several times?
- What is Craig-interpolant of "{A} / {A}" with "A" being unsatisfiable?
- What is Craig-interpolant of "{A, B} / {A}" with "B" being unsatisfiable?

Reply:
After a small fix in the test, every solver agrees on False and True.